design/src/rfcs/rfc0022_refine_builder.md

```diff
@@ -0,0 +1,556 @@
+
+### Cut them down: going from 2N+1 to 2 generic parameters
+
+The previous two examples should have convinced you that the 2N+1 generic parameters on `PokemonServiceBuilder` harm the ergonomics of our API.  
+Can we get rid of them?
+
+Yes! Let's look at one possible approach:
+
+```rust
+pub struct PokemonServiceBuilder<Body, Plugin> {
+    check_health: Option<Route<Body>>,
+    do_nothing: Option<Route<Body>>,
+    get_pokemon_species: Option<Route<Body>>,
+    get_server_statistics: Option<Route<Body>>,
+    capture_pokemon: Option<Route<Body>>,
+    get_storage: Option<Route<Body>>,
+    plugin: Plugin,
+}
+```
+
+We no longer store the raw handlers inside `PokemonServiceBuilder`.
+We eagerly upgrade the operation handlers to a `Route` instance when they are registered with the builder. The existing API performs the upgrade when `build` is called, forcing `PokemonServiceBuilder` to store the raw handlers and keep two generic parameters around (`OpX` and `ExtsX`) for each operation.  
+
+There is one downside to this alternative approach - all plugins must be specified upfront, when creating an instance of the builder.
+Plugins cannot be modified through `PokemonServiceBuilder`'s API if we want to guarantee that all operation handlers are upgraded to a `Route` using the same plugins.
+
+```rust
+impl PokemonService<()> {
+    /// Constructs a builder for [`PokemonService`].
+    pub fn builder<Body, Plugin>(plugin: Plugin) -> PokemonServiceBuilder<Body, Plugin> {
+        PokemonServiceBuilder {
+            check_health: None,
+            do_nothing: None,
+            get_pokemon_species: None,
+            get_server_statistics: None,
+            capture_pokemon: None,
+            get_storage: None,
+            plugin,
+        }
+    }
+}
+```
+
+In practice, this is unlikely to have any impact on developers using `smithy-rs`.  
+We have seen how cumbersome it is to break the startup logic into different functions using the current service builder API. Developers are most likely specifying all plugins and routes in the same function even if the current API allows them to intersperse route registrations and plugin registrations: they would simply have to re-order their registration statements to adopt the API proposed in this RFC.
+
+### Alternatives: boxed trait objects 
+
+Trait objects are another common approach to achieve type erasure: instead of storing the raw handler in `PokemonServiceBuilder`, we box it and cast it to a `dyn Upgradable<...>`.  
+Using trait objects would allow us to delay upgrading to `Route` until `build` is called, mimicking the behaviour of the current API and allowing developers to register plugins at any point before calling `build`.  
+Using trait objects would unfortunately have a non-zero runtime impact: we would be introducing an extra layer of indirection (double-boxing the handler), affecting the performance of our application at runtime.  
```

Comment:
> I don't think this is the case here - we could Box<Box<Box<dyn Upgradable<...>>>> but the actual upgrade procedure would still result in a singly nested handler (Route::new)

Is this the case though?

`Route::new` desugars to a `BoxCloneService::new` call, which introduces one layer of vtable indirection.
The argument to `BoxCloneService::new` is the parameter passed to `Route::new` - if that were to be behind multiple layers of boxing, I am under the impression that we would have to resolve those indirections at runtime when invoking the service.

It's a different situation when we are talking about boxed `Plugin`s - the call to `.map` would suffer from the indirection, but that only happens at startup time.

Have I got this right or am I missing something (more than likely, tbh)?

Reply:
Currently, `Upgradable` looks like this:

```rust
/// Provides an interface to convert a representation of an operation to a HTTP [`Service`](tower::Service) with
/// canonical associated types.
pub trait Upgradable<Protocol, Operation, Exts, B, Plugin> {
    type Service: Service<http::Request<B>, Response = http::Response<BoxBody>>;

    /// Performs an upgrade from a representation of an operation to a HTTP [`Service`](tower::Service).
    fn upgrade(self, plugin: &Plugin) -> Self::Service;
}
```

If we wanted to erase the type we'd have to remove the associated type which would vary across each operation (else we'd have to specify it in the `dyn Upgradable<Service = ....>`). So we'd have to move the `Route::new` into the `Ugpradable`, like this:

```rust
pub trait Upgradable<Protocol, Operation, Exts, B, Plugin> {
    fn upgrade(self, plugin: &Plugin) -> Route<B>;
}
```

I think is acceptable tweak - all the gnarly `Send + Clone + ...` bounds would be pushed down into the `Upgradable for Operation` impl too, rather than hanging around on the service builder which might be worth while anyway (I recall you mentioning this before).

The `.boxed()` method on the service builder would then take the `OpX` and turn it into a `Box<dyn Upgradable<...>>`. The `Box` here would `Box` the `dyn Upgradable` and if we call `.boxed()` multiple times it'd become `Box<Box<Box<dyn Upgradable>>>` but this is `Box`ing the actual `Upgrade` trait. The `Route::new` is still only called once and hence the indirection at runtime only happens once as far as I can tell?